### PR TITLE
Added Upgrade/GetUpgradeStatus/Reboot

### PR DESCRIPTION
**Pull request by cielavenir**
- ~~Not sure if we should open CancelUpgrade()~~
- I suppose we should not open Halt(), although Reboot() is required to restart without autorestart.

### CODE REVIEW

File: include/mujincontrollerclient/mujincontrollerclient.h [hunk 1]
```diff
@@ -499,11 +496,33 @@ class MUJINCLIENT_API ControllerClient
...
     virtual void RestoreBackup(std::istream& inputStream, bool config = true, bool media = true, double timeout = 60.0) = 0;
 
+    /// \brief Upgrade controller's software.
+    ///
+    /// \param inputStream the stream represententing the software file provided by MUJIN.
+    ///        It needs to be seekable to get the size (ifstream subclass is applicable for files).
+    ///        NULL means using previously uploaded file.
+    /// \param autorestart whether to restart automatically after upgrading. if false, Reboot() can be called later.
+    /// \param uploadonly whether to upload only (so that upgrade can be continued later without uploading).
+    virtual void Upgrade(std::istream* inputStream, bool autorestart, bool uploadonly, double timeout = 600.0) = 0;
```
Comment (rdiankov) [inline include/mujincontrollerclient/mujincontrollerclient.h]: @ziyan @kanbouchou @cielavenir

why not just pass in an empty std::istream?
And if its content is empty, then just don't add any form upload, which will just re-use the file.
Or is there a use case where we upgrade from empty file?

Either way, I don't think raw pointer is good here.

Comment (cielavenir) [reply include/mujincontrollerclient/mujincontrollerclient.h]: I tested that (in current client) confirmed that uploading empty stream just causes error on hypervisor. Now I'll change the behavior on empty stream and remove pointer.

Comment (ziyan) [reply include/mujincontrollerclient/mujincontrollerclient.h]: New change looks good to me too. Is it easy to construct an empty istream?

Comment (ziyan) [reply include/mujincontrollerclient/mujincontrollerclient.h]: I see, `stringstream` will do.

**Comment by rdiankov** [comment]
woot, thanks!

**Pull request merged**